File: 81.java
/**
  81. Find Median from Data Stream
  https://www.lintcode.com/problem/find-median-from-data-stream/description
**/
public class Solution {
    /**
     * @param nums: A list of integers
     * @return: the median of numbers
     */
    public int[] medianII(int[] nums) {
        // write your code here
        int n = nums.length;
        PriorityQueue<Integer> max_heap = new PriorityQueue<Integer>(n, Collections.reverseOrder());
        PriorityQueue<Integer> min_heap = new PriorityQueue<Integer>(n);

        int[] results = new int[n];
        for (int i = 0; i < n; ++i) {
            if (max_heap.size() == 0 || nums[i] <= max_heap.peek()) {
                max_heap.offer(nums[i]);
            } else {
                min_heap.offer(nums[i]);
            }
            balance(max_heap, min_heap);

            results[i] = max_heap.peek();
        }

        return results;
    }
    private void balance(PriorityQueue<Integer> max_heap, PriorityQueue<Integer> min_heap) {
        while (min_heap.size() > max_heap.size()) {
            max_heap.offer(min_heap.poll());
        }

        while (max_heap.size() > min_heap.size() + 1) {
            min_heap.offer(max_heap.poll());
        }
    }
}

File: lintcode/81.py
"""
81. Find Median from Data Stream
https://www.lintcode.com/problem/find-median-from-data-stream/description
"""
import heapq
class Solution:
    """
    @param nums: A list of integers
    @return: the median of numbers
    """
    def medianII(self, nums):
        # write your code here
        max_heap = []
        min_heap = []
        results = []
        for num in nums:
            if max_heap and num <= -max_heap[0]:
                heapq.heappush(max_heap, -num)
            else:
                heapq.heappush(min_heap, num)
            self.balance(max_heap, min_heap)

            results.append(-max_heap[0])

        return results

    def balance(self, max_heap, min_heap):
        while len(max_heap) < len(min_heap):
            heapq.heappush(max_heap, -heapq.heappop(min_heap))

        while len(max_heap) > len(min_heap) + 1:
            heapq.heappush(min_heap, -heapq.heappop(max_heap))
